sdk/identity/identity-spa/src/popupCredential.browser.ts

```diff
@@ -0,0 +1,110 @@
+   *
+   * @param options - Options for configuring the client which makes the authentication request.
+   */
+  constructor(options: SPACredentialOptions) {
+    if (!options?.clientId) {
+      const error = new Error(
+        "The parameter `clientId` cannot be left undefined for the `PopupCredential`"
+      );
+      logger.info(formatError("", error));
+      throw error;
+    }
+
+    const msalOptions: MsalBrowserFlowOptions = {
+      ...options,
+      disableAutomaticAuthentication: true,
+      logger,
+      loginStyle: "popup",
+      redirectUri:
+        typeof options.redirectUri === "function" ? options.redirectUri() : options.redirectUri,
+    };
+
+    this.msalFlow = new MSALAuthCode(msalOptions);
+  }
+
+  /**
+   * Authenticates with Azure Active Directory and returns an access token if successful.
+   * If authentication fails, a {@link CredentialUnavailableError} will be thrown with the details of the failure.
+   *
+   * @param scopes - The list of scopes for which the token will have access.
+   * @param options - The options used to configure any requests this
+   *                TokenCredential implementation might make.
+   */
+  async getToken(scopes: string | string[], options: GetTokenOptions = {}): Promise<AccessToken> {
+    return trace(`${this.constructor.name}.getToken`, options, async (newOptions) => {
+      const arrayScopes = Array.isArray(scopes) ? scopes : [scopes];
+      return this.msalFlow.getToken(arrayScopes, {
+        ...newOptions,
+        disableAutomaticAuthentication: true,
+      });
+    });
+  }
+
+  /**
+   * Authenticates with Azure Active Directory and returns an access token if successful.
+   * If authentication fails, a {@link CredentialUnavailableError} will be thrown with the details of the failure.
+   *
+   * If the token can't be retrieved silently, this method will require user interaction to retrieve the token.
+   *
+   * @param scopes - The list of scopes for which the token will have access.
+   * @param options - The options used to configure any requests this
+   *                  TokenCredential implementation might make.
+   */
+  async authenticate(
+    scopes: string | string[],
+    options: GetTokenOptions = {}
+  ): Promise<AuthenticationRecord | undefined> {
+    return trace(`${this.constructor.name}.authenticate`, options, async (newOptions) => {
+      const arrayScopes = Array.isArray(scopes) ? scopes : [scopes];
+      await this.msalFlow.getToken(arrayScopes, newOptions);
+      return this.msalFlow.getActiveAccount();
+    });
+  }
+
+  /**
+   * Retrieves parameters from the URL hash.
+   */
+  async onPageLoad(): Promise<{ code?: string; state?: string }> {
+    const hash = new URLSearchParams(redirectHash);
+    return {
+      code: hash.get("code") || undefined,
+      state: hash.get("state") || undefined,
+    };
+  }
```

Comment:
This isn’t useful for the PopupCredential, but it is useful for the RedirectCredential.